core/src/main/java/software/amazon/awssdk/retry/RetryHandler.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.retry;
+
+import static java.util.Collections.singletonList;
+
+import java.time.Duration;
+import java.util.Optional;
+import software.amazon.awssdk.RequestExecutionContext;
+import software.amazon.awssdk.SdkBaseException;
+import software.amazon.awssdk.http.HttpResponse;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.util.CapacityManager;
+
+public class RetryHandler {
+
+    public static final String HEADER_SDK_RETRY_INFO = "amz-sdk-retry";
+
+    private final RetryPolicy retryPolicy;
+    private final CapacityManager retryCapacity;
+
+    private Duration lastBackoffDelay = Duration.ZERO;
+    private boolean retryCapacityConsumed;
+    private RetryPolicyContext retryPolicyContext;
+    private Optional<SdkBaseException> lastRetriedException = Optional.empty();
+
+    public RetryHandler(RetryPolicy retryPolicy,
+                        CapacityManager retryCapacity) {
+        this.retryPolicy = retryPolicy;
+        this.retryCapacity = retryCapacity;
```

Comment:
Would it be possible for the capacity manager to be part of the retry condition as well? Essentially, you mix-in throttling, the same way we're mixing in the max-retry?

We'd probably need the extra hooks to release the capacity, but it seems weird that an SDK customer can't implement retry throttling using our condition interface.

Reply:
It could but I'm not sure it's worth the extra complexity. This is always the first condition that should be checked as you'd be wasting time if it wasn't. We'd also have to help a lot through our builder as the customer shouldn't need to know which exceptions are throttling if they want to override the defaults.

Reply:
The complexity argument is legit - I'm not sure how much extra it would be. Would it be more than 1 method?

The wasting-time argument I don't follow so much. What's another few nanoseconds added to the case that the throttle condition will fail?

In terms of the customer not knowing about which exceptions are throttling exceptions, the same thing applies to the customer not knowing which exceptions are retryable exceptions. If that's a problem here, it's a problem there as well.

Reply:
> In terms of the customer not knowing about which exceptions are throttling exceptions, the same thing applies to the customer not knowing which exceptions are retryable exceptions.

I don't think that's equivalent. In the retryable exception case, there is one dimension, exceptions. They either take ours or set their own.

In a condition dealing with throttling they are more likely interested in modifying the number of throttled retries and the cost while using our throttled retries. In that case they need to know what our throttled exceptions are to be able to produce a complete retry condition.

> The wasting-time argument I don't follow so much. What's another few nanoseconds added to the case that the throttle condition will fail?

Ultimately I guess it could apply in any situation where you have a set of conditions but I would typically see a priority to retry conditions in which throttling is the highest. Evaluating any condition before it would be a waste of time.

Reply:
It just feels like our retry policy interface isn't powerful enough to express what we need, so we're working around it to add extra features. It feels like LESS complexity to have it under just one abstraction. I suppose it's always something we could add later, assuming we don't expose the concept of capacity manager outside of the core.

Would a reasonable compromise be to hide all traces of it from our public interfaces?

Reply:
I'm on the fence. I see where you're coming from Matt and I agree it would make sense in the retry condition. We would need some way to notify the retry condition that a request has succeeded though and I feel like that would pollute the interface a bit (no longer a simple predicate). Not exactly sure how many customers would actually want to implement their own throttling so I'd say lean towards leaving it out for now.